File: from-0601-to-0700/0647-palindromic-substrings/solution.go
func countSubstrings(s string) int {
    n := len(s)
    if n < 2 {
        return n
    }

    count := n
    dp := make([][]int, n)
    for i := 0; i < n; i ++ {
        dp[i] = make([]int, n)
        dp[i][i] = 1
    }

    for col := 1; col < n; col ++ {
        for row := 0; row < col; row ++ {
            if s[row] == s[col]{
                if row == col - 1 || dp[row + 1][col - 1] == 1{
                    dp[row][col] = 1
                    count += 1
                }
            }
        }
    } 
    return count
}

File: from-0601-to-0700/0647-palindromic-substrings/solution.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        slen = len(s)
        if slen < 2:
            return slen
        
        count = 0
        dp = [ [0] * slen for _ in range(slen)]
        for i in range(slen):
            dp[i][i] = 1
            count += 1
            
        for col in range(1, slen):
            for row in range(col):
                if s[row] == s[col]:                # the first and the last chars are equal             
                    if row == col - 1:              # the len of substr == 2      
                        dp[row][col] = 1
                        count += 1
                    elif dp[row+1][col-1] == 1:     # the inner substr is a palindrome
                        dp[row][col] = 1
                        count += 1
                    
        return count
                    
            